pkg/controller/storagecluster/initialization_reconciler.go

```diff
@@ -55,9 +70,118 @@ func (r *ReconcileStorageCluster) ensureStorageClasses(instance *ocsv1.StorageCl
+	if instance.Status.ExternalSecretFound {
+		return nil
+	}
+	found := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      externalClusterDetailsSecret,
+			Namespace: instance.Namespace,
+		},
+	}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: found.Name, Namespace: found.Namespace}, found)
+	if err != nil {
+		return err
+	}
+	ownerRef := metav1.OwnerReference{
+		UID:        instance.UID,
+		APIVersion: instance.APIVersion,
+		Kind:       instance.Kind,
+		Name:       instance.Name,
+	}
+	var data []ocsv1.ExternalResource
+	err = json.Unmarshal(found.Data[externalClusterDetailsKey], &data)
+	if err != nil {
+		return fmt.Errorf("could not parse json blob, got %+v", err)
+	}
+	// storageClasses stores all the StorageClasses coming from the JSON
+	var storageClasses map[string]map[string]string = make(map[string]map[string]string)
+
+	for _, d := range data {
+		objectMeta := metav1.ObjectMeta{
+			Name:            d.Name,
+			Namespace:       instance.Namespace,
+			OwnerReferences: []metav1.OwnerReference{ownerRef},
+		}
+		objectKey := types.NamespacedName{Name: d.Name, Namespace: instance.Namespace}
+		switch d.Kind {
+		case "ConfigMap":
+			cm := &corev1.ConfigMap{
+				ObjectMeta: objectMeta,
+				Data:       d.Data,
+			}
+			found := &corev1.ConfigMap{ObjectMeta: objectMeta}
+			err := r.client.Get(context.TODO(), objectKey, found)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					reqLogger.Info("Creating Ceph ConfigMap:	" + cm.Name)
+					err = r.client.Create(context.TODO(), cm)
+					if err != nil {
+						return err
+					}
+				}
+				return err
+			}
+		case "Secret":
+			s := &corev1.Secret{
+				ObjectMeta: objectMeta,
+				Data:       make(map[string][]byte),
+			}
+			for k, v := range d.Data {
+				s.Data[k] = []byte(v)
+			}
+			found := &corev1.Secret{ObjectMeta: objectMeta}
+			err := r.client.Get(context.TODO(), objectKey, found)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					reqLogger.Info("Creating Ceph Secret: 	" + s.Name)
+					err = r.client.Create(context.TODO(), s)
+					if err != nil {
+						return err
+					}
+				}
+				return err
+			}
+		case "StorageClass":
+			storageClasses[d.Name] = d.Data
+		}
+	}
+	scs, err := r.newStorageClasses(instance)
+	if err != nil {
+		return err
+	}
+	if len(scs) < 3 {
+		scs = append(scs, r.newOBCStorageClass(instance))
+	}
+	var testStorageClass *storagev1.StorageClass
+	for scName, scData := range storageClasses { //iterating over the map
+		if scName == "cephfs" {
+			// Setting the fsname for cephfs StorageClass
+			testStorageClass = scs[0]
+			// Setting the PoolName cephfs StorageClass
+		} else if scName == "ceph-rbd" {
+			// Setting the PoolName for RBD StorageClass
+			testStorageClass = scs[1]
+		} else if scName == "ceph-rgw" {
+			// Setting the Endpoint for OBC StorageClass
+			testStorageClass = scs[2]
+		}
+		for k, v := range scData {
+			testStorageClass.Parameters[k] = v
+		}
+	}
```

Comment:
Why can't you do this in the `case "StorageClass":` above?

Reply:
**Please ask clarification questions if needed.**
The reason is that whenever we call `newStorageClass` it will return us 3 Storage classes.
So once we will have all the StorageClass information at one place, then we can process them And start modifying them.

I hope that makes sense :)